pkg/endpoint/policy.go

```diff
@@ -143,79 +150,143 @@ func (e *Endpoint) setNextPolicyRevision(revision uint64) {
+	err = e.rlockAlive()
+	if err != nil {
+		return nil, err
+	}
 
 	// No point in calculating policy if endpoint does not have an identity yet.
 	if e.SecurityIdentity == nil {
 		e.getLogger().Warn("Endpoint lacks identity, skipping policy calculation")
-		return nil
+		e.runlock()
+		return nil, nil
 	}
 
+	// Copy out some values we care about, then unlock
+	forcePolicyCompute := e.forcePolicyCompute
+	epNextPolicyRevision := e.nextPolicyRevision
+	epPolicyRevision := e.policyRevision
+	selectorPolicy := e.selectorPolicy
+	securityIdentity := e.SecurityIdentity
+	e.runlock()
+
 	e.getLogger().Debug("Starting policy recalculation...")
 	stats := &policyRegenerationStatistics{}
 	stats.totalTime.Start()
+	defer func() {
+		stats.totalTime.End(err == nil)
+		e.updatePolicyRegenerationStatistics(stats, forcePolicyCompute, err)
+	}()
 
 	stats.waitingForPolicyRepository.Start()
 	repo := e.policyGetter.GetPolicyRepository()
 	repo.Mutex.RLock()
-	revision := repo.GetRevision()
-	defer repo.Mutex.RUnlock()
+	repoRevision := repo.GetRevision()
 	stats.waitingForPolicyRepository.End(true)
 
 	// Recompute policy for this endpoint only if not already done for this revision.
-	if !e.forcePolicyCompute && e.nextPolicyRevision >= revision {
+	if !forcePolicyCompute && epNextPolicyRevision >= repoRevision && e.desiredPolicy != nil {
 		e.getLogger().WithFields(logrus.Fields{
-			"policyRevision.next": e.nextPolicyRevision,
-			"policyRevision.repo": revision,
-			"policyChanged":       e.nextPolicyRevision > e.policyRevision,
+			"policyRevision.next": epNextPolicyRevision,
+			"policyRevision.repo": repoRevision,
+			"policyChanged":       epNextPolicyRevision > epPolicyRevision,
 		}).Debug("Skipping unnecessary endpoint policy recalculation")
-
-		return nil
+		repo.Mutex.RUnlock()
+		return nil, nil
 	}
 
 	stats.policyCalculation.Start()
-	if e.selectorPolicy == nil {
+	if selectorPolicy == nil {
 		// Upon initial insertion or restore, there's currently no good
 		// trigger point to ensure that the security Identity is
 		// assigned after the endpoint is added to the endpointmanager
 		// (and hence also the identitymanager). In that case, detect
 		// that the selectorPolicy is not set and find it.
-		e.selectorPolicy = repo.GetPolicyCache().Lookup(e.SecurityIdentity)
-		if e.selectorPolicy == nil {
+		//nolint:staticcheck
+		selectorPolicy = repo.GetPolicyCache().Lookup(securityIdentity)
+		//nolint:staticcheck // linter thinks that Lookup cannot return nil, it can.
+		if selectorPolicy == nil {
 			err := fmt.Errorf("no cached selectorPolicy found")
 			e.getLogger().WithError(err).Warning("Failed to regenerate from cached policy")
-			return err
+			repo.Mutex.RUnlock()
+			return nil, err
 		}
 	}
-	// TODO: GH-7515: This should be triggered closer to policy change
-	// handlers, but for now let's just update it here.
-	if err := repo.GetPolicyCache().UpdatePolicy(e.SecurityIdentity); err != nil {
```

Comment:
This todo is not actually resolved. The todo is basically saying that policy computation can be done outside of the endpoint completely. While this overall PR gets us closer in that direction, we are technically still in the context of the endpoint. 

Going forward, we can resolve #7515 when we can calculate policy for endpoints (outside of the context of a single endpoint) when
1. policies change
2. identities change
3. endpoints change

and "change" here means added or removed. 

Concretely what this could look like is some `func CalculatePolicy(identity *identity.Identity)` which we call on each of the above listed occurrences. Then the only logic necessary in the endpoint should just be a lookup like 

```
selectorPolicy := repo.PolicyCache().Lookup(e.SecurityIdentity)
...
if forcePolicyCompute {
  repo.PolicyCache().UpdatePolicy(e.SecurityIdentity)
}
```

I toyed with this in https://github.com/squeed/cilium/compare/policy-allocate-when-used-v2...christarazi:pr/christarazi/separate-policy-calc-ep-regen, but obviously needs more work. Just wanted to share more context.

Reply:
aha, I saw that the issue was closed and didn't really go much further than that.

Not to get too distracted from the current issue, but I'm not sure that just extracting policy computation is that interesting. More useful would be the ability to fully push policy to bpf & Envoy outside of a regeneration. I don't know how interleaved they are, but it seems like it could be doable.

Reply:
> Not to get too distracted from the current issue, but I'm not sure that just extracting policy computation is that interesting. More useful would be the ability to fully push policy to bpf & Envoy outside of a regeneration. I don't know how interleaved they are, but it seems like it could be doable.

Agreed, and I think we need to extract / decouple policy calculation from endpoint regeneration in order to be able to push policy to BPF & Envoy outside of a regeneration, in order to simplify the model. Today, the logic is a bit interleaved and harder to reason about. 

A regeneration is considered a full policy calculation, while `ApplyPolicyMapChanges` is incremental. If we separate policy calculation completely from endpoint, then that should in theory allow us to _always_ perform incremental policy updates, which should improve performance, overhead, and reduce potential interleaving of critical sections. It will also put the code into a structure where it's easier to introduce `pkg/stream` infrastructure.

The way I see it is we can have a model where `CalculatePolicy(identity)` returns a policy revision and the desired computed policy, and hands that to each endpoint that has `identity`. Each endpoint with that `identity` will then take their realized, current policy and compute the diff between the desired policy from `CalculatePolicy`, and implements the diff. The endpoints would then be considered implementing the returned policy revision.

In contrast to today, each endpoint attempts to compute the policy diff. `N` endpoints -> compute policies `N` times (there's some caching of course). 

We sort of "invert" this relationship with `CalculatePolicy` by doing: Policy computed once -> `N` endpoints updated. 

This makes the model much simpler to think about and should enable us to always have incremental policy updates.

Reply:
All of this is a fine idea - and would indeed be cleaner - except for the specter of named ports. This is basically what forces policy calculation to be done in two phases, an identity-generic phase and and an endpoint-specific phase.

Ultimately, we don't currently detect the "edge" that is an Identity being assigned to an Endpoint on the node. If we build such a thing in to the EndpointManager, then that seems like the right place to move some of these pure control-plane computations.

Reply:
This reminds me of an aspect of named ports being resolved per-endpoint after a shared `selectorPolicy` has been computed: It is possible for a named port resolve to the same port that is used in the policy by number, and in the extreme case it would be possible for both of them specify L7 policies for different proxy parsers. Which one gets plumbed to the datapath/Envoy is a matter of map iteration order. Otherwise I think we handle this gracefully, e.g.:
- the policy with a deny rule gets to the datapath, if both are denies that is not a problem
- if one is redirect while other is not, the one with redirect takes precedence

Reply:
> All of this is a fine idea - and would indeed be cleaner - except for the specter of named ports. This is basically what forces policy calculation to be done in two phases, an identity-generic phase and and an endpoint-specific phase.

Agreed, good point. It would be really nice to make that two-pronged policy computation (identity-based & named port-based) very clear in the code. I think it's very doable.  

> Ultimately, we don't currently detect the "edge" that is an Identity being assigned to an Endpoint on the node. If we build such a thing in to the EndpointManager, then that seems like the right place to move some of these pure control-plane computations.

Yep, we can definitely take steps to improve that. :+1: